### ISSUE TITLE
multi-gpu with forward hook when training supernet with resnet

### ISSUE THREAD

**Issue by deJQK** [mit-han-lab/gan-compression#33]
This is a great project. However, I met the similar issue as [this](https://github.com/mit-han-lab/gan-compression/issues/27). Moreover, the issue seems to be a little randomly, such that training with 2 GPUs can sometimes run correctly but sometimes not. The key point is that the intermediate features are obtained from forward hook and stored in dictionaries, which cannot guarantee correct device as the netA. The bug occurs [here](https://github.com/mit-han-lab/gan-compression/blob/master/supernets/resnet_supernet.py#L63-L65), where netA is always on cuda:0 because it is not wrapped with data parallel correctly (https://github.com/mit-han-lab/gan-compression/blob/master/distillers/base_resnet_distiller.py#L117-L123 , where it should be netA = networks.init_net(netA, gpu_ids=self.gpu_ids), and there should not be to(device) in the above two cases; [this line](https://github.com/mit-han-lab/gan-compression/blob/master/supernets/resnet_supernet.py#L61) and [this line](https://github.com/mit-han-lab/gan-compression/blob/master/supernets/resnet_supernet.py#L65) should be modified accordingly to something like getattr(netA, 'module', netA)). However, since Tact and Sact are randomly on different devices, and dictionary will lose this information, there will still be some bug on this. One possible solution is to include such device in the keys [here](https://github.com/mit-han-lab/gan-compression/blob/master/distillers/base_resnet_distiller.py#L159) to include information of device of output, but since netA is data parallel, which scatter inputs and replicate weights during its forward, netA(Sact) still does not work. If moving netA to the device of Sact before calling it, the optimizerG step will cause problem as the grad will be on different device as the netA's weight. I am not sure if there is a solution on this. I notice you changed the code structure in the spade net, but I wonder if there can be simpler solution.

**Comment by lmxyy** [comment]
Yes, you are right. The training of SPADE always requires multiple GPUs, so I changed the hook-based intermediate feature obtaining to a forward-based one. Since I only train pix2pix and cycleGAN on a single GPU, so I didn't fix this bug. I will try to fix it. Thanks for pointing it out!

**Comment by deJQK** [comment]
Thanks @lmxyy for the reply. Do you think there can be a simpler solution beyond changing the code structure, for example, by moving netA to the device of Sact, and gather the gradient to the device of netA so that the weights are updated correctly?

**Comment by deJQK** [comment]
Hi @lmxyy , I got one solution here. Hope it is helpful.

First change [this line](https://github.com/mit-han-lab/gan-compression/blob/master/distillers/base_resnet_distiller.py#L159) to
`mem[name+str(output.device)] = output`
and you can remove [this line](https://github.com/mit-han-lab/gan-compression/blob/master/distillers/base_resnet_distiller.py#L123) as it is not used.

Inside the [resnet_supernet.py](https://github.com/mit-han-lab/gan-compression/blob/master/supernets/resnet_supernet.py) file, you need first import the following:
`from torch.nn.parallel.data_parallel import gather, replicate, parallel_apply`
and change [these lines](https://github.com/mit-han-lab/gan-compression/blob/master/supernets/resnet_supernet.py#L63-L66) to
netA_replicas = replicate(netA, self.gpu_ids)
kwargs = tuple([{'config': {'channel': netA.out_channels}} for idx in self.gpu_ids])
Sacts = parallel_apply(netA_replicas, tuple([self.Sacts[key] for key in sorted(self.Sacts.keys()) if n in key]), kwargs)
Tacts = [self.Tacts[key] for key in sorted(self.Tacts.keys()) if n in key]

**Comment by deJQK** [comment]
Hi @lmxyy , I got one solution here. Hope it is helpful.

First change [this line](https://github.com/mit-han-lab/gan-compression/blob/master/distillers/base_resnet_distiller.py#L159) to
`mem[name+str(output.device)] = output`
and you can remove [this line](https://github.com/mit-han-lab/gan-compression/blob/master/distillers/base_resnet_distiller.py#L123) as it is not used.

Inside the resnet_supernet.py file, you need first import the following:

`from torch.nn.parallel.data_parallel import gather, replicate, parallel_apply`

and change [these lines](https://github.com/mit-han-lab/gan-compression/blob/master/supernets/resnet_supernet.py#L63-L66) to

```
netA_replicas = replicate(netA, self.gpu_ids)
kwargs = tuple([{'config': {'channel': netA.out_channels}} for idx in self.gpu_ids])
Sacts = parallel_apply(netA_replicas, tuple([self.Sacts[key] for key in sorted(self.Sacts.keys()) if n in key]), kwargs)
Tacts = [self.Tacts[key] for key in sorted(self.Tacts.keys()) if n in key]
loss = [F.mse_loss(Sact, Tact) for Sact, Tact in zip(Sacts, Tacts)]
loss = gather(loss, self.gpu_ids[0]).sum()
```

**Comment by MuyangLi-Dawnlight** [comment]
Thanks, I've fixed the bug with your suggestions.

**Comment by lmxyy** [comment]
Ok, I've fix the bug with your suggestions. Thanks for pointing it out!
